### PR TITLE
Stability improvements and debugging support for storage layer1 replication.

### PR DESCRIPTION
**Pull request by andyHa**

### CODE REVIEW

File: src/main/java/sirius/biz/analytics/scheduler/BaseEntityBatchEmitter.java [hunk 1]
```diff
@@ -118,7 +118,7 @@
 
         Class<E> type = (Class<E>) mixing.getDescriptor(typeName).getType();
         Q query = (Q) getMapper().select(type);
-        query.where(getMapper().filters().gte(BaseEntity.ID, startId))
+        query.where(getMapper().filters().gt(BaseEntity.ID, startId))
```
Comment (idlira) [inline src/main/java/sirius/biz/analytics/scheduler/BaseEntityBatchEmitter.java:121]: shouldn't it be inverted: >= startId and < endId ?

File: src/main/java/sirius/biz/storage/s3/ObjectStore.java [hunk 2]
```diff
@@ -307,7 +307,7 @@ public void listObjects(BucketName bucket, @Nullable String prefix, Predicate<S3
         TaskContext taskContext = CallContext.getCurrent().get(TaskContext.class);
 
         do {
-            try (Operation op = new Operation(() -> Strings.apply("Fetching S3 objects from %s (prefix: %s)",
+            try (Operation op = new Operation(() -> Strings.apply("S3: Fetching objects from %s (prefix: %s)",
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/s3/ObjectStore.java:310]: Optional: Rename `op` to `operation`

File: src/main/java/sirius/biz/storage/layer1/replication/jdbc/SQLReplicationTaskStorage.java [hunk 3]
```diff
@@ -42,6 +42,8 @@
      */
     public static final String FRAMEWORK_JDBC_REPLICATION = "biz.storage-replication-jdbc";
 
+    private static final String TXN_ID = "txnId";
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer1/replication/jdbc/SQLReplicationTaskStorage.java:45]: Optional: Rename `TXN_ID` to `TRANSACTION_ID`. I am not really a fan of the shortened field name but at least name the constant more descriptive please

File: src/main/java/sirius/biz/storage/layer1/replication/jdbc/SQLReplicationTaskStorage.java [hunk 4]
```diff
@@ -110,52 +109,67 @@ public void notifyAboutUpdate(String primarySpace, String objectId, long content
...
-                               batch -> {
-                                   distributedTasks.submitFIFOTask(ReplicationTaskExecutor.class, batch);
-                                   scheduledTasks.addAndGet(batchSize);
 
-                                   markTasksAsScheduled(batch);
-
-                                   return numberOfBatches.decrementAndGet() > 0;
-                               });
+        while (numberOfBatches.decrementAndGet() > 0) {
+            String txnId = Strings.generateCode(32);
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer1/replication/jdbc/SQLReplicationTaskStorage.java:114]: Optional: Rename to transactionId

File: src/main/java/sirius/biz/storage/layer1/replication/jdbc/SQLReplicationTask.java [hunk 5]
```diff
@@ -72,6 +72,14 @@
     @NullAllowed
     private LocalDateTime scheduled;
 
+    /**
+     * Contains the batch or transaction identifier used to schedule this task.
+     */
+    public static final Mapping TXN_IN = Mapping.named("txnId");
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer1/replication/jdbc/SQLReplicationTask.java:78]: Typo: `TXN_IN` to `TXN_ID`

File: src/main/java/sirius/biz/storage/layer1/replication/jdbc/SQLReplicationTaskStorage.java [hunk 6]
```diff
@@ -110,52 +109,67 @@ public void notifyAboutUpdate(String primarySpace, String objectId, long content
...
-            query.eq(SQLReplicationTask.SCHEDULED, null);
-        } else {
-            query.ne(SQLReplicationTask.SCHEDULED, null);
-        }
-    }
-
     @Override
     public void executeBatch(JSONObject batch) {
-        emitter.evaluateBatch(batch, query -> filterOnExecutableTasks(query, false), this::executeTask);
+        String txnId = batch.getString(TXN_ID);
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer1/replication/jdbc/SQLReplicationTaskStorage.java:166]: Optional: s.o.

File: src/main/java/sirius/biz/storage/layer1/replication/mongo/MongoReplicationTaskStorage.java [hunk 7]
```diff
@@ -43,6 +44,8 @@
      */
     public static final String FRAMEWORK_MONGO_REPLICATION = "biz.storage-replication-mongo";
 
+    private static final String TXN_ID = "txnId";
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer1/replication/mongo/MongoReplicationTaskStorage.java:47]: Optional: s.o.

**Review by sabieber — APPROVED** [review]

**Review by mkeckmkeck — APPROVED** [review]

File: src/main/java/sirius/biz/analytics/scheduler/BaseEntityBatchEmitter.java (hunk 1 above)
Comment (andyHa) [reply src/main/java/sirius/biz/analytics/scheduler/BaseEntityBatchEmitter.java:121]: nope - should be good (??)

Comment (idlira) [reply src/main/java/sirius/biz/analytics/scheduler/BaseEntityBatchEmitter.java:121]: In the first run, how do we reach the first Id?

File: src/main/java/sirius/biz/storage/layer1/replication/jdbc/SQLReplicationTaskStorage.java [hunk 8]
```diff
@@ -111,17 +111,17 @@ public int emitBatches() {
...
-                markTaskAsScheduled(task, txnId);
+                markTaskAsScheduled(task, transactionId);
                 tasksFound.set(true);
                 scheduledTasks.incrementAndGet();
             });
 
             if (tasksFound.get()) {
                 distributedTasks.submitFIFOTask(ReplicationTaskExecutor.class,
-                                                new JSONObject().fluentPut(TXN_ID, txnId));
+                                                new JSONObject().fluentPut(TXN_ID, transactionId));
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer1/replication/jdbc/SQLReplicationTaskStorage.java:124]: Should the JSON property also be renamed to transactionId?

**Review by sabieber — APPROVED** [review]

File: src/main/java/sirius/biz/analytics/scheduler/BaseEntityBatchEmitter.java (hunk 1 above)
Comment (andyHa) [reply src/main/java/sirius/biz/analytics/scheduler/BaseEntityBatchEmitter.java:121]: in the first run startId is null or "" and thus the constraint is ignored...

**Review by idlira — APPROVED** [review]

**Review by idlira — APPROVED** [review]

**Review by sabieber — APPROVED** [review]

**Review by ymo-sci — APPROVED** [review]

**Review by cschierle — APPROVED** [review]

**Pull request merged**